app/commands/user_track/generate_summary.rb

```diff
@@ -2,75 +2,189 @@ class UserTrack
   class GenerateSummary
     include Mandate
 
-    initialize_with :user_track
-    delegate :track, to: :user_track
+    initialize_with :track, :user_track
 
+    # We want to ensure we only do this once, but we
+    # provide two access vectors, this and the
+    # user_track.summary. This slightly hacky method
+    # checks whether we've already done this work and
+    # if so returns what we already have. Otherwise
+    # it does the work and then sets it.
+    #
+    # If we don't have a user_track, then we just generate
+    # the data and don't worry about this bit.
     def call
-      #####################
-      # Populate concepts #
-      #####################
-      concepts = {}
-      track.concepts.each do |concept|
-        concept_exercises = concept_exercise_data.values.
+      return generate! unless user_track
+
+      return user_track.summary if user_track.summary_generated?
+
+      user_track.summary = generate!
+    end
+
+    def generate!
+      t = Time.now.to_f
+      Rails.logger.info "[BM] Starting generating user summary"
+
+      UserTrack::Summary.new(
+        concepts.map { |k, v| [k, ConceptSummary.new(v)] }.to_h,
+        exercises.map { |k, v| [k, ExerciseSummary.new(v)] }.to_h
+      ).tap do
```

Comment:
We're doing this mapping here (rather than lower down) because we'll cache JSON, not the serialized objects, so this method will eventually have a "cache or generate stuff - then map to the structs"